### ISSUE TITLE
Incremental HTML building broken

### ISSUE THREAD

**Issue by dopplershift** [sphinx-doc/sphinx#3774]
Subject: Second build without clean (i.e. `make html`) fails with:
```pytb
# Sphinx version: 1.6.1
# Python version: 3.6.1 (CPython)
# Docutils version: 0.13.1 release
# Jinja2 version: 2.9.5
# Last messages:
#   building [mo]: targets for 0 po files that are out of date
#   
#   building [html]: targets for 0 source files that are out of date
#   
#   updating environment:
#   
#   0 added, 19 changed, 0 removed
#   
#   reading sources... [  5%] api/generated/metpy.calc.density
#   
# Loaded extensions:
#   alabaster (0.7.10) from /Users/rmay/miniconda3/envs/py36/lib/python3.6/site-packages/alabaster/__init__.py
#   sphinx.ext.autodoc (1.6.1) from /Users/rmay/miniconda3/envs/py36/lib/python3.6/site-packages/sphinx/ext/autodoc.py
#   sphinx.ext.autosummary (1.6.1) from /Users/rmay/miniconda3/envs/py36/lib/python3.6/site-packages/sphinx/ext/autosummary/__init__.py
#   sphinx.ext.coverage (1.6.1) from /Users/rmay/miniconda3/envs/py36/lib/python3.6/site-packages/sphinx/ext/coverage.py
#   sphinx.ext.intersphinx (1.6.1) from /Users/rmay/miniconda3/envs/py36/lib/python3.6/site-packages/sphinx/ext/intersphinx.py
#   sphinx.ext.mathjax (1.6.1) from /Users/rmay/miniconda3/envs/py36/lib/python3.6/site-packages/sphinx/ext/mathjax.py
#   sphinx.ext.napoleon (1.6.1) from /Users/rmay/miniconda3/envs/py36/lib/python3.6/site-packages/sphinx/ext/napoleon/__init__.py
#   sphinx.ext.viewcode (1.6.1) from /Users/rmay/miniconda3/envs/py36/lib/python3.6/site-packages/sphinx/ext/viewcode.py
#   matplotlib.sphinxext.plot_directive (unknown version) from /Users/rmay/miniconda3/envs/py36/lib/python3.6/site-packages/matplotlib/sphinxext/plot_directive.py
Traceback (most recent call last):
  File "/Users/rmay/miniconda3/envs/py36/lib/python3.6/site-packages/sphinx/cmdline.py", line 306, in main
    app.build(opts.force_all, filenames)
  File "/Users/rmay/miniconda3/envs/py36/lib/python3.6/site-packages/sphinx/application.py", line 338, in build
    self.builder.build_update()
  File "/Users/rmay/miniconda3/envs/py36/lib/python3.6/site-packages/sphinx/builders/__init__.py", line 328, in build_update
    'out of date' % len(to_build))
  File "/Users/rmay/miniconda3/envs/py36/lib/python3.6/site-packages/sphinx/builders/__init__.py", line 341, in build
    updated_docnames = set(self.env.update(self.config, self.srcdir, self.doctreedir))
  File "/Users/rmay/miniconda3/envs/py36/lib/python3.6/site-packages/sphinx/environment/__init__.py", line 584, in update
    self._read_serial(docnames, self.app)
  File "/Users/rmay/miniconda3/envs/py36/lib/python3.6/site-packages/sphinx/environment/__init__.py", line 602, in _read_serial
    self.clear_doc(docname)
  File "/Users/rmay/miniconda3/envs/py36/lib/python3.6/site-packages/sphinx/environment/__init__.py", line 323, in clear_doc
    domain.clear_doc(docname)
  File "/Users/rmay/miniconda3/envs/py36/lib/python3.6/site-packages/sphinx/domains/std.py", line 542, in clear_doc
    docnames.pop(docname)
TypeError: 'str' object cannot be interpreted as an integer
```
### Problem
I'm trying to use sphinx-gallery together with sphinx's link check capability. This fails with the above message, but I'm readily able to reproduce without either.

#### Procedure to reproduce the problem
1. `make clean html`
2. Edit a python file (NOT just an rst file)
3. `make html`

#### Error logs / results
First build works:
```
rm -rf build/* ./examples ./tutorials ./api/generated
sphinx-autogen -i -t ./_templates -o ./api/generated ./api/*.rst
[autosummary] generating autosummary for: ./api/index.rst
[autosummary] writing to ./api/generated
[autosummary] generating autosummary for: ./api/generated/metpy.calc.rst, ./api/generated/metpy.constants.rst, ./api/generated/metpy.gridding.rst, ./api/generated/metpy.io.cdm.rst, ./api/generated/metpy.io.rst, ./api/generated/metpy.plots.ctables.rst, ./api/generated/metpy.plots.rst, ./api/generated/metpy.units.rst
[autosummary] writing to ./api/generated
[autosummary] generating autosummary for: ./api/generated/metpy.calc.advection.rst, ./api/generated/metpy.calc.convergence_vorticity.rst, ./api/generated/metpy.calc.coriolis_parameter.rst, ./api/generated/metpy.calc.density.rst, ./api/generated/metpy.calc.dewpoint.rst, ./api/generated/metpy.calc.dewpoint_rh.rst, ./api/generated/metpy.calc.dry_lapse.rst, ./api/generated/metpy.calc.el.rst, ./api/generated/metpy.calc.equivalent_potential_temperature.rst, ./api/generated/metpy.calc.find_intersections.rst, ..., ./api/generated/metpy.plots.ctables.read_colortable.rst, ./api/generated/metpy.units.DimensionalityError.rst, ./api/generated/metpy.units.PintConverter.rst, ./api/generated/metpy.units.UndefinedUnitError.rst, ./api/generated/metpy.units.atleast_1d.rst, ./api/generated/metpy.units.atleast_2d.rst, ./api/generated/metpy.units.check_units.rst, ./api/generated/metpy.units.concatenate.rst, ./api/generated/metpy.units.iterable.rst, ./api/generated/metpy.units.masked_array.rst
[autosummary] writing to ./api/generated
sphinx-build -b html -d build/doctrees   . build/html
Running Sphinx v1.6.1
making output directory...
loading pickled environment... not yet created
loading intersphinx inventory from http://matplotlib.org/objects.inv...
loading intersphinx inventory from https://docs.python.org/3/objects.inv...
loading intersphinx inventory from https://docs.scipy.org/doc/numpy/objects.inv...
loading intersphinx inventory from https://docs.scipy.org/doc/scipy/reference/objects.inv...
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 93 source files that are out of date
updating environment: 93 added, 0 changed, 0 removed
reading sources... [100%] units                                                                                                        
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] units                                                                                                         
generating indices... genindex py-modindex
highlighting module code... [100%] matplotlib.cbook                                                                                    
writing additional pages... search
copying images... [100%] _static/radar.png                                                                                             
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded.

Build finished. The HTML pages are in build/html.
```
Second build (incremental, without `clean`):
```
>make html
sphinx-autogen -i -t ./_templates -o ./api/generated ./api/*.rst
[autosummary] generating autosummary for: ./api/index.rst
[autosummary] writing to ./api/generated
sphinx-build -b html -d build/doctrees   . build/html
Running Sphinx v1.6.1
loading pickled environment... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 0 source files that are out of date
updating environment: 0 added, 19 changed, 0 removed

Exception occurred:
  File "/Users/rmay/miniconda3/envs/py36/lib/python3.6/site-packages/sphinx/domains/std.py", line 542, in clear_doc
    docnames.pop(docname)
TypeError: 'str' object cannot be interpreted as an integer
The full traceback has been saved in /var/folders/r8/bhrhn4m56lgdckj491rm6_w40000gp/T/sphinx-err-yret87iy.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
make: *** [html] Error 1
```

#### Expected results
I expect the second `make html` to successfully rebuild files that changed. This is necessary so that `linkcheck` can run properly after sphinx-gallery has done some updating of files, but it seems pretty important to have incremental building working.

### Reproducible project / your project
- github.com/Unidata/MetPy
- https://travis-ci.org/Unidata/MetPy/jobs/234183545

### Environment info
- OS: OSX 10.12 and Ubuntu 12.04 (Travis) can both reproduce
- Python version: 3.6
- Sphinx version: 1.6.1

**Comment by dopplershift** [comment]
For my particular failure above, I put a print in the sphinx code for `docnames` and `docname` and got:

Docnames: ['api/generated/metpy.calc.density', 'api/generated/metpy.calc.virtual_temperature'] Docname: api/generated/metpy.calc.density        

Seems like at that point someone is treating a list like a dictionary.

**Comment by tk0miya** [comment]
Sorry for inconvenience.
Fixed at 16d0569.

**Comment by dopplershift** [comment]
Thanks for the quick fix!
